veps/sig-compute/hypervisor-feature-discovery.md

```diff
@@ -0,0 +1,372 @@
+        WithMessage("VGA graphics not supported on ARM64 architecture").
+        WithDocLink("https://kubevirt.io/user-guide/graphics#compatibility").
+        Done()
+    
+    // VGA is unregistered on amd64/s390x - implicitly allowed
+}
+```
+
+### Resolution Strategy
+
+The registry resolves capabilities using a layered fallback strategy, checking from most specific to least specific:
+
+```go
+// Query with automatic fallback: hypervisor/arch -> hypervisor -> base
+func Get(hypervisor, arch string, cap CapabilityKey) Capability {
+    // Try exact match: kvm/amd64
+    if c, ok := lookup(hypervisor+"/"+arch, cap); ok { return c }
+    
+    // Try hypervisor-wide: kvm
+    if c, ok := lookup(hypervisor, cap); ok { return c }
+    
+    // Not registered - return zero value (caller checks cap.Level)
+    return Capability{}
+}
+```
+
+Resolution precedence (most specific wins):
+1. **Hypervisor+Architecture** (e.g., `kvm/amd64`) - most specific
+2. **Hypervisor-wide** (e.g., `kvm`) - applies to all architectures unless overridden
+3. **Not registered** - returns zero-value `Capability{}` (where `Level` is 0, which equals `Unregistered`)
+
+### Integration Points
+
+#### Validation Webhooks
+
+The validating webhook integrates with the capability registry by:
+
+1. Mapping VMI spec features to capability keys (e.g., `video.type: vga` → `graphics.vga`)
+2. Querying the registry for each capability
+3. Applying validation logic based on the returned `Capability` data (support level, message, doc links)
+
+The registry provides the data; consumers apply their own validation policy. This pattern replaces scattered validation files (`arm64.go`, `s390x.go`) with centralized data queries.
+
+
+#### Test Filtering
+
+Tests use capability decorators to automatically skip on unsupported platforms:
+
+```go
+// tests/decorators/capabilities.go
+var RequiresVGAGraphics = decorators.RequiresCapability(capabilities.CapGraphicsVGA)
+
+It("should support VGA graphics", RequiresVGAGraphics, func() {
+    vmi := libvmifact.NewCirros()
+    vmi.Spec.Domain.Devices.Video = &v1.Video{Type: "vga"}
+})
+```
+
+Behavior:
+- `Unregistered`: test runs (default for unregistered capabilities)
+- `Unsupported`: test skips with message from registry
+- `Experimental`: test runs only if feature gate enabled
+- `Deprecated`: test runs with warning (feature still works)
+
+Example: VGA tests run on KVM/amd64 (unregistered, implicitly allowed) but skip on KVM/arm64 (registered as unsupported).
+
+#### Node Labeling
+
+The node labeller can optionally expose capability labels on nodes for scheduling and audit purposes:
+
+```go
+// Example node labels generated from capability registry (only for registered capabilities)
+kubevirt.io/capability.graphics.vga: "unsupported"  // Only on arm64
+kubevirt.io/capability.cpu.hotplug: "experimental"
+```
+
+This allows users and operators to query node capabilities without inspecting the hypervisor configuration.
```

Comment:
Do node labels have precedence over the capability matrix?

Also, if the node labeller already sets those according to the architecture of the node, do we even need to include architecture in the matrix? Having the hypervisor as the only possible key would simplify it...

Reply:
Agreed. The validation aspect of this VEP is focused on validating VMI spec at admission-time, where architecture information would be necessary.

I am not sure of the benefit of labeling nodes with VMI-feature level capabilities supported on its architecture. For the VMI to enter `Scheduling` state, we can assume that its required capabilities have been successfully validated against the capability registry. The same registry's information was also used to generate the node labels. Therefore using capability-based node selectors for VMI scheduling would be redundant.